angularSTEP/src/app/upload-recipe/upload-recipe.component.ts

```diff
@@ -0,0 +1,106 @@
+import { Component } from '@angular/core';
+import { FormControl, FormGroup, Validators } from '@angular/forms';
+import { AngularFirestore } from '@angular/fire/firestore';
+import { Router } from '@angular/router';
+import { formattedError } from '@angular/compiler';
+
+@Component({
+  selector: 'app-upload-recipe',
+  templateUrl: './upload-recipe.component.html',
+  styleUrls: ['./upload-recipe.component.scss'],
+})
+
+export class UploadRecipeComponent {
+  constructor(private db: AngularFirestore,
+              private router: Router,) {
+  }
+
+  ingredients: string[] = [];
+  instructions: string[] = [];
+  tools: string[] = [];
+
+  recipeForm = new FormGroup({
+    name: new FormControl('', [Validators.required]),
+    description: new FormControl('', [Validators.required]),
+    tools: new FormControl(this.tools, [Validators.required]),
+    ingredients: new FormControl(this.ingredients, [Validators.required]),
+    instructions: new FormControl(this.instructions, [Validators.required]),
+    extraInfo: new FormControl(''),
+    difficulty: new FormControl('', [Validators.required]),
+  });
+
+
+  addIngredient(newIngredient: string) {
+    var formattedIngredient: string = this.autoCapitalizeFirst(newIngredient);
+    if (formattedIngredient != '' && formattedIngredient != undefined) {
+      this.ingredients.push(formattedIngredient);
```

Comment:
I'm concerned about injection attacks along the lines of adding an ingredient/instruction like `<script>alert('oops');</script>`. Does Angular escape these automatically when adding them to the DOM?

Reply:
I read into this and Angular does have some built in security features. In this case the ingredient/instruction/tool added to the DOM is done so through interpolation, which Angular always automatically escapes. In the case of something like what you suggested being entered, the resulting HTML would not be interpreted and would be displayed in the list with angle brackets as text.

Reply:
Awesome!